### PR TITLE
Add ShapeOfConstFolding transform

### PR DESCRIPTION
**Pull request by iimironov**
Ticket: 52451

Description:
Transformation which transform: Const() - ShapeOf() -> Const() subgraph. 

* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR
* [x]  Transformation preserves original framework node names N/A: a special case, when we should rename the nodes

Validation:
* [x]  Unit tests
* [x]  Framework operation tests N/A
* [x]  Transformation tests N/A
* [x]  Model Optimizer IR Reader check 

Documentation:
* [x]  Supported frameworks operations list N/A: no new operations were enabled
* [x]  Guide on how to convert the **public** model N/A no new models were enabled
* [x]  User guide update N/A

### CODE REVIEW

File: model-optimizer/extensions/middle/ShapeOfConstFolding.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    """
+    TThe transformation calculates the value of the constant
+    Const() -> ShapeOf() --> Const()
+    """
+    enabled = True
+
+    def pattern(self):
+        return dict(
+            nodes=[
+                ('const', dict(op='Const')),
```
Comment (jane-intel) [inline model-optimizer/extensions/middle/ShapeOfConstFolding.py:24]: It's middle replacer. So pattern should have datas

**Review by jane-intel — CHANGES_REQUESTED** [review]
Transformation doesn't work

**Comment by jane-intel** [comment]
Please add template description to PR

File: model-optimizer/extensions/middle/ShapeOfConstFolding.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+                ('const', 'const_data', {}),
+                ('const_data', 'shape', {})
+            ])
+
+    def replace_pattern(self, graph: Graph, match: [dict, SubgraphMatch]):
+        shape = match['shape']
+
+        shape_name = shape.soft_get('name', shape.id)
+
+        shape_value = shape.in_port(0).data.get_value()
```
Comment (jane-intel) [inline model-optimizer/extensions/middle/ShapeOfConstFolding.py:34]: the code doesn't work again. it is the shape that should become the value, not the constant value.

File: model-optimizer/extensions/middle/ShapeOfConstFolding.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+                ('shape', dict(op='ShapeOf')),
+            ],
+            edges=[
+                ('const', 'const_data', {}),
+                ('const_data', 'shape', {})
+            ])
+
+    def replace_pattern(self, graph: Graph, match: [dict, SubgraphMatch]):
+        shape = match['shape']
+        if shape.has_valid('value'):
```
Comment (jane-intel) [inline model-optimizer/extensions/middle/ShapeOfConstFolding.py:31]: I assume `ShapeOf` op node doesn't have value. Only data nodes have them.

File: model-optimizer/extensions/back/ShapeOfConstFolding.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    The transformation folding Const() -> ShapeOf() to Const()
+    """
+    enabled = True
+
+    def run_after(self):
+        from extensions.back.MatMulNormalizer import SmartReshape_HC_Reshape_MatMul
+        return [SmartReshape_HC_Reshape_MatMul]
+
+    def find_and_replace_pattern(self, graph: Graph):
+        print("ShapeOfConstFolding")
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/back/ShapeOfConstFolding.py:20]: I think that this debug code should be removed. The same comment for line 22.

File: model-optimizer/unit_tests/extensions/back/ShapeOfConstFolding_test.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+import unittest
+
+import numpy as np
+
+from extensions.back.ShapeOfConstFolding import ShapeOfConstFolding
+from mo.front.common.partial_infer.utils import int64_array
+from mo.utils.ir_engine.compare_graphs import compare_graphs
+from unit_tests.utils.graph import build_graph
```
Comment (lazarevevgeny) [inline model-optimizer/unit_tests/extensions/back/ShapeOfConstFolding_test.py:11]: sort imports

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: model-optimizer/extensions/back/ShapeOfConstFolding.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+from mo.back.replacement import BackReplacementPattern
+from mo.graph.graph import Graph, rename_nodes
+from mo.ops.const import Const
+
+
+class ShapeOfConstFolding(BackReplacementPattern):
+    """
+    The transformation folding Const() -> ShapeOf() to Const()
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/back/ShapeOfConstFolding.py:11]: `The transformation folding Const() -> ShapeOf() to Const()` -> `The transformation folds ShapeOf(Const) -> Const`.

File: model-optimizer/unit_tests/extensions/back/ShapeOfConstFolding_test.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+                    # new nodes
+                    'new_const_shapeof': {'type': 'Const', 'kind': 'op', 'op': 'Const',
+                                          'value': int64_array([1, 3, 30, 30])},
+                    'new_const_shapeof_data': {'value': int64_array([1, 3, 30, 30]), 'kind': 'data'},
+                    }
+
+
+class ShapeOfConstFoldingTests(unittest.TestCase):
+    def test1(self):
```
Comment (lazarevevgeny) [inline model-optimizer/unit_tests/extensions/back/ShapeOfConstFolding_test.py:40]: I don't see a test case the for following graph:

image

File: model-optimizer/unit_tests/extensions/back/ShapeOfConstFolding_test.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+                                 ('new_const_shapeof_data', 'mul'),
+                                 ('mul', 'mul_data'),
+                                 ('mul_data', 'last')],
+                                {
+                                    'input': {'shape': int64_array([1, 3, 30, 30])},
+                                    'input_data': {'shape': int64_array([1, 3, 30, 30])},
+                                    'shapeof_input': {'value': int64_array([1, 3, 30, 30])},
+                                    'shapeof_input_data': {'value': int64_array([1, 3, 30, 30])},
+                                    'new_const_shapeof': {'value': int64_array([1, 3, 30, 30])},
+                                    'new_const_shapeof_data': {'value': int64_array([1, 3, 30, 30])},
```
Comment (lazarevevgeny) [inline model-optimizer/unit_tests/extensions/back/ShapeOfConstFolding_test.py:80]: Why the graph contains new data node `new_const_shapeof_data`? The `shapeof_const_data` data node should be kept! Looks like you loose the debug information (tensor name) after running this transformation.

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: model-optimizer/unit_tests/extensions/back/ShapeOfConstFolding_test.py [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+                                    'new_const_shapeof': {'value': int64_array([1, 3, 30, 30])},
+                                    'shapeof_const_data': {'value': int64_array([1, 3, 30, 30])},
+                                    'mul_data': {'value': int64_array([1, 9, 900, 900])},
+                                },
+                                nodes_with_edges_only=True)
+        ShapeOfConstFolding().find_and_replace_pattern(graph)
+        (flag, resp) = compare_graphs(graph, graph_ref, 'last')
+        self.assertTrue(flag, resp)
+
+    def test_const_with_tow_outputs(self):
```
Comment (mvafin) [inline model-optimizer/unit_tests/extensions/back/ShapeOfConstFolding_test.py:118]: Typo: tow -> two

**Review by mvafin — APPROVED** [review]

File: model-optimizer/unit_tests/extensions/back/ShapeOfConstFolding_test.py [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+from extensions.back.ShapeOfConstFolding import ShapeOfConstFolding
+from mo.front.common.partial_infer.eltwise import eltwise_infer
+from mo.front.common.partial_infer.utils import int64_array
+from mo.utils.ir_engine.compare_graphs import compare_graphs
+from unit_tests.utils.graph import build_graph
+
+const_value = np.random.rand(1, 3, 30, 30)
+nodes_attributes = {'input': {'shape': int64_array([1, 3, 30, 30]), 'type': 'Parameter', 'kind': 'op',
+                              'op': 'Parameter'},
+                    'input_data': {'value': None, 'shape': int64_array([1, 3, 30, 30]), 'kind': 'data'},
```
Comment (mvafin) [inline model-optimizer/unit_tests/extensions/back/ShapeOfConstFolding_test.py:17]: nit: use helper functions for graph generation to create data nodes and connect edges.

**Review by achetver — APPROVED** [review]

**Review by lazarevevgeny — APPROVED** [review]
@jane-intel , please, review today.

**Review by jane-intel — APPROVED** [review]

**Pull request merged**